### DISCUSSION TITLE
Clipboard does not copy/paste objects when complied with .Net8

### DISCUSSION THREAD

**Discussion by cornerbowlsoftware** [AvaloniaUI/Avalonia#16350] [General]
### Describe the bug

Clipboard does not copy/paste objects when complied with .Net8

### To Reproduce

Create a new Avalonia project.
Set the targets to: 
`<TargetFramework>net7.0</TargetFramework>`
Create a model that includes the [Serializable] attribute. For example:

```
    [Serializable]
    public class MyTestModel
    {
        public int Id { get; set; }
        public string? Name { get; set; }

        public MyTestModel() 
        { 
        }
    }
```
Update the MainView to create the model and copy then paste the model to and from the clipboard like so:

```
public partial class MainView : UserControl
{
    public MainView()
    {
        InitializeComponent();
        Loaded += MainView_Loaded;
    }

    private void MainView_Loaded(object? sender, Avalonia.Interactivity.RoutedEventArgs e)
    {
        TestAsync().GetAwaiter().GetResult();
    }

    private async Task TestAsync()
    {
        var m = new MyTestModel()
        {
            Id = 1,
            Name = "Test",
        };

        var dataObject = new DataObject();
        dataObject.Set("my-app-person", m);

        var clipboard = TopLevel.GetTopLevel(this)!.Clipboard!;
        await clipboard.SetDataObjectAsync(dataObject);

        var r = await clipboard.GetDataAsync("my-app-person");
        Debug.WriteLine(r);
    }
}
```
Run the program. Notice the variable "r" is returned as expected.
Modify both Avalonia projects to target: 
`<TargetFramework>net8.0</TargetFramework>`
Run the program. Notice the variable "r" is null.

### Expected behavior

The value should be returned.

### Avalonia version

11.0.11

### OS

Windows

**Comment by timunie** [comment]
Please don't do GetAwaiter.getresult. I told you before. Instead make OnLoaded async void (this is allowed) and call the function via await

**Comment by kekekeks** [comment]
IIRC BinaryFormatter got deprecated or disabled with net8.0. Do not rely on the old .NET binary serialization when using modern .NET runtimes

**Comment by maxkatz6** [comment]
Yes, it's not going to be supported anymore.
We might provide some plug-in data-object serializer/deserializer. But generally, you should do it on your app side passing byte array to the API.

**Comment by cornerbowlsoftware** [comment]
@kekekeks and @maxkatz6 my WPF apps work as expected.  This bug appears to be specific to Avalonia.

Thanks.

**Comment by timunie** [comment]
Please upload a minimum sample using latest nightly builds. Don't use GetAwaiter in the test project. Thanks.

**Comment by cornerbowlsoftware** [comment]
@timunie please see the following sample application that exposes the bug.

[AvaloniaClipboardBug.zip](https://github.com/user-attachments/files/16258516/AvaloniaClipboardBug.zip)

**Comment by maxkatz6** [comment]
@cornerbowlsoftware 

> This bug appears to be specific to Avalonia.

FYI, https://github.com/dotnet/announcements/issues/293
Also https://learn.microsoft.com/en-us/dotnet/core/compatibility/serialization/8.0/binaryformatter-disabled

**Reply by cornerbowlsoftware** [reply]
@maxkatz6 Yes, I already saw that, however, my .Net8 WPF app works. The bug is specific to Avalonia.  There is something else going on.

**Reply by IanRawley** [reply]
>Previous behavior
>
>In .NET 7, the [BinaryFormatter.Serialize(Stream, Object)](https://learn.microsoft.com/en-us/dotnet/api/>system.runtime.serialization.formatters.binary.binaryformatter.serialize#system-runtime-serialization-formatters-binary->binaryformatter-serialize(system-io-stream-system-object)) and [BinaryFormatter.Deserialize(Stream)](https://learn.microsoft.com/en->us/dotnet/api/system.runtime.serialization.formatters.binary.binaryformatter.deserialize#system-runtime-serialization-formatters->binary-binaryformatter-deserialize(system-io-stream)) methods were marked obsolete and raised an error at compile time. However, if your application suppressed the obsoletion, it could still call the methods and they functioned properly in most project types (excluding >ASP.NET, WASM, and MAUI). For example, the APIs functioned correctly in a console app.

> New behavior
>
>Starting in .NET 8, the affected methods throw a [NotSupportedException](https://learn.microsoft.com/en-us/dotnet/api/>system.notsupportedexception) at run time **across all project types except Windows Forms and WPF.** The APIs continue to remain >obsolete (as error) across all project types, including Windows Forms and WPF.

From the second link.  This is not a bug with Avalonia, it's WPF behaving differently to everything else.

**Reply by cornerbowlsoftware** [reply]
@IanRawley ok so you are saying even though .Net 8 Windows and 4.7.2 WPF are able to do it, the underlying platform specific libraries in Avalonia are unable to utilize the functionality .Net 8 Windows is using?  Ok.  I don't need inter-application copy paste so it was super easy to simply use a singleton to save my objects for later pasting.

**Reply by timunie** [reply]
@cornerbowlsoftware yep. That's what the NET team did. Disabled it for anyone except WPF and WinForms. We cannot change that from our side. But there's a switch for you to enable it again.

**Reply by kekekeks** [reply]
Again, it's not a bug, it's a critical security vulnerability that got fixed by Microsoft.
You can re-enable the formatter by adding  `<EnableUnsafeBinaryFormatterSerialization>true</EnableUnsafeBinaryFormatterSerialization>` to your project file if you wish to voluntarily subject your users to a risk, but we won't be setting this flag from Avalonia package

**Comment by kekekeks** [comment]
Not a bug, people who don't care about security can re-enable the feature [as documented here](https://github.com/dotnet/designs/blob/main/accepted/2020/better-obsoletion/binaryformatter-obsoletion.md#apis-previously-marked-obsolete-as-warning-in-net-5-now-produce-build-errors-by-default-net-7).

Note that the API will be removed from .NET 9 entirely, so fix your apps.

**Reply by cornerbowlsoftware** [reply]
@timunie and @kekekeks Ok thanks for letting me know.  I already moved away and just save the copied object to a singleton.  I'll do the same in my WPF app.  Thanks for the heads up.
